Changes in module 'nf-core/concoct/cutupfasta'
--- modules/nf-core/concoct/cutupfasta/main.nf
+++ modules/nf-core/concoct/cutupfasta/main.nf
@@ -1,12 +1,11 @@
-
 process CONCOCT_CUTUPFASTA {
     tag "$meta.id"
     label 'process_single'
 
     conda "bioconda::concoct=1.1.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/concoct:1.1.0--py311h245ed52_4':
-        'biocontainers/concoct:1.1.0--py311h245ed52_4' }"
+        'https://depot.galaxyproject.org/singularity/concoct:1.1.0--py38h7be5676_2':
+        'biocontainers/concoct:1.1.0--py38h7be5676_2' }"
 
     input:
     tuple val(meta), path(fasta)
@@ -37,4 +36,4 @@
         concoct: \$(echo \$(concoct --version 2>&1) | sed 's/concoct //g' )
     END_VERSIONS
     """
-}
+}
************************************************************
