Changes in module 'nf-core/maxbin2'
--- modules/nf-core/maxbin2/main.nf
+++ modules/nf-core/maxbin2/main.nf
@@ -1,5 +1,5 @@
 process MAXBIN2 {
-    tag "$meta.id"
+    tag "$meta.id $contigs"
     label 'process_medium'
 
     conda "bioconda::maxbin2=2.2.7"
@@ -11,7 +11,7 @@
     tuple val(meta), path(contigs), path(reads), path(abund)
 
     output:
-    tuple val(meta), path("*.fasta.gz")   , emit: binned_fastas
+    tuple val(meta), path("*.fasta")   , emit: binned_fastas
     tuple val(meta), path("*.summary")    , emit: summary
     tuple val(meta), path("*.log.gz")     , emit: log
     tuple val(meta), path("*.marker.gz")  , emit: marker_counts
@@ -23,6 +23,16 @@
 
     when:
     task.ext.when == null || task.ext.when
+
+    errorStrategy {
+        task.exitStatus {
+            exitVal ->
+                // Retry on non-zero exit codes
+                return exitVal != 0 ? ErrorAction.RETRY : ErrorAction.FINISH
+        }
+        maxRetries 3  // Set the maximum number of retries
+        sleep 10       // Set the delay between retries in seconds
+    }
 
     script:
     def args = task.ext.args ?: ''
@@ -37,7 +47,7 @@
         $args \\
         -out $prefix
 
-    gzip *.fasta *.noclass *.tooshort *log *.marker
+    gzip *.noclass *.tooshort *log *.marker
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
